cvat/apps/dataset_manager/bindings.py

```diff
@@ -2151,10 +2151,66 @@ def reduce_fn(acc, v):
                 raise CvatImportError("Image {}: can't import annotation "
                     "#{} ({}): {}".format(item.id, idx, ann.type.name, e)) from e
 
-    for track in tracks.values():
-        track['elements'] = list(track['elements'].values())
-        instance_data.add_track(instance_data.Track(**track))
+    def _validate_track_shapes(shapes):
+        new_shapes = []
+        prev_shape = None
+        # infer the keyframe shapes and keep only them
+        for shape in shapes:
+            prev_is_visible = prev_shape and not prev_shape.outside
+            cur_is_visible = shape and not shape.outside
+            if not cur_is_visible:
+                if prev_is_visible:
+                    shape = shape._replace(keyframe=True)
+                    new_shapes.append(shape)
+                prev_shape = shape
+                continue
+
+            has_gap = False
+            if prev_is_visible:
+                has_gap = prev_shape.frame + instance_data.frame_step < shape.frame
+
+            if not prev_is_visible or has_gap or shape.keyframe:
+                if has_gap:
+                    prev_shape = prev_shape._replace(outside=True, keyframe=True,
+                        frame=prev_shape.frame + instance_data.frame_step)
+                    new_shapes.append(prev_shape)
 
+                shape = shape._replace(keyframe=True)
+                new_shapes.append(shape)
+
+            prev_shape = shape
```

Comment:
```suggestion
            prev_is_visible = prev_shape and not prev_shape.outside
            cur_is_visible = shape and not shape.outside

            has_gap = False
            if prev_is_visible:
                has_gap = prev_shape.frame + instance_data.frame_step < shape.frame

            if has_gap:
                prev_shape = prev_shape._replace(outside=True, keyframe=True,
                    frame=prev_shape.frame + instance_data.frame_step)
                new_shapes.append(prev_shape)

            if prev_is_visible != cur_is_visible or cur_is_visible and (has_gap or shape.keyframe):
                shape = shape._replace(keyframe=True)
                new_shapes.append(shape)

            prev_shape = shape
```

Seems like this should fix the problem

Reply:
Yeah, this makes the logic clearer and allows for handling all the cases.

I am currently thinking of how to test this [case](https://github.com/cvat-ai/cvat/pull/7669#discussion_r1570266503). Using our current test to test the case of two consecutive outside attributes might be hard. This is because when we export the annotation, we will ignore the second `outside` = `True`.


"|" - shape
"*" - outside true
":" - no shape

before export
[ | | | * : : * : : | | | | ]

After export it out:
[ | | | : : : : : : | | | | ] (it will skip outside=True [here](https://github.com/cvat-ai/cvat/blob/a5269b056341db726da01c2d779d6e3dbb20c8bc/cvat/apps/dataset_manager/bindings.py#L436)

Import the annotation:
[ | | | * : : : : : | | | | ]

So, the imported annotation will only contain one outside instead of two.
This case is more likely to handle human-crafted or annotations created by other tools.

Do you have any idea about testing this case?

Reply:
Actually, the example above wasn't manually edited, it's just a CVAT export in COCO keypoints for the skeleton track described (3 points, 6 frames, 1st frame visible, 2nd and 3rd fully outside, then points appear by one consequently). So it goes like this: `[ | * : | | |  ]`

Reply:
Sorry for the confusion made. I did not notice the example when writing the message. Will take a look now.